### PR TITLE
call livenessCheck periodically

### PR DESCRIPTION
**Pull request by xzfc**
## Description

Check data-plane aliveness periodically, not only after a control-plane event happens

Continuation of PR #1252.

## Issue link

#1187

## How Has This Been Tested?

- [x] Added unit testing to cover
- [x] Tested manually
- [ ] Tested by integration testing
- [ ] Have not tested

## Types of changes

- [ ] Bug fix
- [x] New functionallity
- [ ] Documentation
- [ ] Refactoring
- [ ] CI

### CODE REVIEW

File: pkg/networkservice/common/heal/client.go [hunk 1]
```diff
@@ -31,21 +32,26 @@ import (
...
+	livenessCheckInterval time.Duration
 }
 
 // NewClient - returns a new heal client chain element
 func NewClient(chainCtx context.Context, opts ...Option) networkservice.NetworkServiceClient {
-	o := &option{
-		livelinessCheck: func(_ *networkservice.Connection) bool { return false },
+	o := &options{
+		attemptAfter:          time.Millisecond * 200,
+		livenessCheckInterval: time.Millisecond * 500,
```
Comment (edwarnicke) [inline pkg/networkservice/common/heal/client.go:45]: Out of curiosity, why these default values?

Comment (xzfc) [reply pkg/networkservice/common/heal/client.go]: * `attemptAfter` is the same as before this PR.

* About `livenessCheckInterval`. In the latest revision of the PR, this value was removed; the user should provide its own timer instead in the `LivenessChecker` implementation.

  In the previous version of this PR (the one that you commented), `livenessCheckInterval` had roughly the same order of magnitude as the default `ping` utility interval. Also, it was tied to [these values](https://github.com/Chunosov/cmd-nsc/blob/6fca644fa41fe5a8d533375c088a1b010d8b7ee6/main.go#L74).

File: pkg/networkservice/common/heal/eventloop.go [hunk 2]
```diff
@@ -32,18 +31,19 @@ import (
 	"github.com/networkservicemesh/sdk/pkg/tools/log"
 )
 
-const attemptAfter = time.Millisecond * 200
+const maxAttemptsWithoutReselect = 3
```
Comment (denis-tingaikin) [inline pkg/networkservice/common/heal/eventloop.go:34]: @xzfc Could you say more why do we need this?

Comment (xzfc) [reply pkg/networkservice/common/heal/eventloop.go:34]: I've added it as a workaround to [this issue](https://github.com/networkservicemesh/sdk/issues/1187#issuecomment-1002990951) (NSE pod was deleted, but it still was responding to pings), not knowing that this was fixed in https://github.com/networkservicemesh/sdk-vpp/pull/537.

However, I think it still makes sense to perform a forceful reselect after a few failed requests to the same endpoint, even if the data plane is still alive.

Comment (denis-tingaikin) [reply pkg/networkservice/common/heal/eventloop.go:34]: We should keep requests without reselect while datapath is fine.

Comment (xzfc) [reply pkg/networkservice/common/heal/eventloop.go]: Removed `maxAttemptsWithoutReselect`.

File: pkg/networkservice/common/heal/client.go (hunk 1 above)
Comment (edwarnicke) [reply pkg/networkservice/common/heal/client.go]: @xzfc Are we using the 'ping' command, or are we sending the icmp echo messages directly from Go?

Comment (xzfc) [reply pkg/networkservice/common/heal/client.go]: We're sending ICMP messages directly, see https://github.com/networkservicemesh/sdk-kernel/pull/447.

Comment (edwarnicke) [reply pkg/networkservice/common/heal/client.go]: @xzfc Woot!

File: pkg/networkservice/common/heal/eventloop.go [hunk 3]
```diff
@@ -78,55 +76,113 @@ func newEventLoop(ctx context.Context, cc grpc.ClientConnInterface, conn *networ
...
+	return res
+}
+
+func (cev *eventLoop) eventLoop() {
+	reselect := false
+
+	ctrlPlaneCh := cev.waitCtrlPlaneEvent()
+
+	livenessCheckerCtx, livenessCheckerCancel := context.WithCancel(context.Background())
+	dataPlaneCh := make(chan struct{}, 1)
```
Comment (denis-tingaikin) [inline pkg/networkservice/common/heal/eventloop.go:136]:
```suggestion
```

File: pkg/networkservice/common/heal/eventloop.go [hunk 4]
```diff
@@ -78,55 +76,113 @@ func newEventLoop(ctx context.Context, cc grpc.ClientConnInterface, conn *networ
...
+	defer livenessCheckerCancel()
+	if cev.heal.livenessChecker != nil {
+		go func() {
+			cev.heal.livenessChecker(livenessCheckerCtx, cev.conn)
+			dataPlaneCh <- struct{}{}
```
Comment (denis-tingaikin) [inline pkg/networkservice/common/heal/eventloop.go:141]:
```suggestion
			livenessCheckerCancel()
```

File: pkg/networkservice/common/heal/client.go [hunk 5]
```diff
@@ -32,20 +33,22 @@ import (
 
 type healClient struct {
 	chainCtx        context.Context
-	livelinessCheck LivelinessCheck
+	livenessChecker LivenessChecker
+	attemptAfter    time.Duration
```
Comment (denis-tingaikin) [inline pkg/networkservice/common/heal/client.go:37]: Why do we need this?

File: pkg/networkservice/common/heal/client.go [hunk 6]
```diff
@@ -32,20 +33,22 @@ import (
...
 }
 
 // NewClient - returns a new heal client chain element
 func NewClient(chainCtx context.Context, opts ...Option) networkservice.NetworkServiceClient {
-	o := &option{
-		livelinessCheck: func(_ *networkservice.Connection) bool { return false },
+	o := &options{
+		attemptAfter: time.Millisecond * 200,
```
Comment (denis-tingaikin) [inline pkg/networkservice/common/heal/client.go:43]:
```suggestion
```

File: pkg/networkservice/common/heal/client.go [hunk 7]
```diff
@@ -32,20 +33,22 @@ import (
...
 	}
 	for _, opt := range opts {
 		opt(o)
 	}
 	return &healClient{
 		chainCtx:        chainCtx,
-		livelinessCheck: o.livelinessCheck,
+		livenessChecker: o.livenessChecker,
+		attemptAfter:    o.attemptAfter,
```
Comment (denis-tingaikin) [inline pkg/networkservice/common/heal/client.go:51]:
```suggestion
```

File: pkg/networkservice/common/heal/eventloop.go [hunk 8]
```diff
@@ -78,55 +76,113 @@ func newEventLoop(ctx context.Context, cc grpc.ClientConnInterface, conn *networ
...
+
 	/* Attempts to heal the connection */
-	afterTicker := clock.FromContext(cev.eventLoopCtx).Ticker(attemptAfter)
+	afterTicker := clock.FromContext(cev.eventLoopCtx).Ticker(cev.heal.attemptAfter)
 	defer afterTicker.Stop()
 	for {
 		select {
 		case <-cev.chainCtx.Done():
 			return
+		case <-dataPlaneCh:
```
Comment (denis-tingaikin) [inline pkg/networkservice/common/heal/eventloop.go:173]: This confuses. Sometimes it calls dataPlane, sometimes it calls as livenesscheker. Can we follow one style naming?

File: pkg/networkservice/common/heal/eventloop.go [hunk 9]
```diff
@@ -78,55 +76,113 @@ func newEventLoop(ctx context.Context, cc grpc.ClientConnInterface, conn *networ
...
+			cev.logger.Warnf("Data plane is down")
+			reselect = true
 		case <-afterTicker.C():
```
Comment (denis-tingaikin) [inline pkg/networkservice/common/heal/eventloop.go:176]: What if request try was too long and the ticker channel is not empty?

File: pkg/networkservice/common/heal/eventloop.go [hunk 10]
```diff
@@ -78,55 +76,113 @@ func newEventLoop(ctx context.Context, cc grpc.ClientConnInterface, conn *networ
...
 }
 
-func (cev *eventLoop) eventLoop() {
-	/* Receive monitor events */
-	for {
-		eventIn, err := cev.client.Recv()
-		if cev.eventLoopCtx.Err() != nil {
-			return
-		}
+func (cev *eventLoop) waitCtrlPlaneEvent() <-chan struct{} {
```
Comment (denis-tingaikin) [inline pkg/networkservice/common/heal/eventloop.go:95]: consider better naming

File: pkg/networkservice/common/heal/options.go [hunk 11]
```diff
@@ -17,22 +17,33 @@
...
+	"context"
+	"time"
+
 	"github.com/networkservicemesh/api/pkg/api/networkservice"
 )
 
-// LivelinessCheck - function that returns true of conn is 'live' and false otherwise
-type LivelinessCheck func(conn *networkservice.Connection) bool
+// LivenessChecker blocks until the connection is alive or ctx is done.
+type LivenessChecker func(ctx context.Context, conn *networkservice.Connection)
```
Comment (denis-tingaikin) [inline pkg/networkservice/common/heal/options.go:27]: @edwarnicke Is this API looking good for you?

**Review by denis-tingaikin — COMMENTED** [review]
In general looks OK. Added a few comments related to style and pointed at potential attemptAfter leak

File: pkg/networkservice/common/heal/client.go (hunk 5 above)
Comment (xzfc) [reply pkg/networkservice/common/heal/client.go]: To make sandbox tests run faster. In the latest revision of this PR replaced with the `BackOff` interface.

File: pkg/networkservice/common/heal/eventloop.go (hunk 8 above)
Comment (xzfc) [reply pkg/networkservice/common/heal/eventloop.go]: Renamed `livenessChecker` to `dataPlaneLivenessChecker`.

File: pkg/networkservice/common/heal/eventloop.go (hunk 9 above)
Comment (xzfc) [reply pkg/networkservice/common/heal/eventloop.go]: Added `cev.chainCtx.Err() != nil` check. Also replaced `case <-dataPlaneCh` with `dataPlaneCtx.Err() != nil` check.

File: pkg/networkservice/common/heal/eventloop.go (hunk 10 above)
Comment (xzfc) [reply pkg/networkservice/common/heal/eventloop.go]: Renamed to `monitorCtrlPlane()`.

**Comment by glazychev-art** [comment]
See - https://github.com/networkservicemesh/sdk/pull/1271

**Pull request closed without merging**